arch/x64/arch-elf.hh

```diff
@@ -42,4 +42,37 @@ enum {
 
 #define ELF_KERNEL_MACHINE_TYPE 62
 
+#define SAFETY_BUFFER 256
+#include <osv/align.hh>
+
+inline void elf_entry_point(void* ep, int argc, char** argv, int argv_size)
+{
+    //The layout of the stack and state of all relevant registers is described
+    //in detail in the section 3.4 (Process Initialization) of the System V Application
+    //Binary Interface AMD64 Architecture Processor Supplement Draft Version 0.95
+    //(see https://refspecs.linuxfoundation.org/elf/x86_64-abi-0.95.pdf)
+    int argc_plus_argv_stack_size = argv_size + 1;
+
+    void *stack;
+    asm volatile ("movq %%rsp, %0" : "=r"(stack));
+    //
+    //According to the document above the stack pointer should be 16-bytes aligned
+    stack -= (SAFETY_BUFFER + argc_plus_argv_stack_size * sizeof(char*));
+    stack = align_down(stack, 16);
+
+    //... and it should start with argc, followed by argv, environment pointers and
+    //auxiliary vector entries. For details look at application::prepare_argv()
+    *reinterpret_cast<u64*>(stack) = argc;
+    memcpy(stack + sizeof(char*), argv, argv_size * sizeof(char*));
+
+    //TODO: Reset SSE2 and floating point registers and RFLAGS
```

Comment:
Why do we need to reset anything regarding the floating point? Won't the executable do whatever it wants with it?

Reply:
Because this is what the spec https://refspecs.linuxfoundation.org/elf/x86_64-abi-0.95.pdf says:
"3.4 Process Initialization
3.4.1 Initial Stack and Register State
Special Registers
The AMD64 architecture defines floating point instructions. At process startup
the two floating point units, SSE2 and x87, both have all floating-point exception
status flags cleared. The status of the control words is as defined in tables 3.3 and
3.4.
Table 3.3: x87 Floating-Point Control Word
Field Value Note
RC 0 Round to nearest
PC 11 Double extended precision
PM 1 Precision masked
UM 1 Underflow masked
OM 1 Overflow masked
ZM 1 Zero divide masked
DM 1 Denormal operand masked
IM 1 Invalid operation masked
"

Reply:
I put better comments.